### PR TITLE
header utility: support removing dot segment in the path header before matching

### LINKED ISSUE #6008: How to normalize URL in Http::HeaderMap?
*Title*: *How to normalize URL in Http::HeaderMap?*

*Description*:
If a URL path includes double dots (e.g. `/data/../info/`) or single dot (e.g. `/user/./record/`), it seems `Http::HeaderMap` doesn't normalize it by removing the dots in the path.

Sometime this creates issues when a filter is trying to compare the path. For example, a filter may expect to match on path `/info/` and it actually wants to match on both `/data/../info/` and `/info/` because they have the same semantics and will most likely to be interpreted to the same place in the backend.

*Question*:
- Should we provide such support in `HeaderMap` or `HeaderUtility::matchHeaders`? Or should we simply do it inside a specific filter when needed?

*Reference*: 
https://en.wikipedia.org/wiki/URL_normalization

/cc @liminw @JimmyCYJ

### PR DESCRIPTION
**Pull request by yangminzhu**
Signed-off-by: Yangmin Zhu <ymzhu@google.com>

**Description**: this PR adds a `MatchOption` to `HeaderUtility::matchHeaders()` to allow removing the dot segments in the path header before matching.
**Risk Level**: low
**Testing**: unit tests.
**Docs Changes**: updated the header matcher comment in rbac API
**Release Notes**: added for the behavior change in rbac filter
Fixes #6008
[Optional Deprecated:]

The only normalization we need for now is to remove the dots from the path.
/cc @lizan @htuch @liminw

### CODE REVIEW

File: source/common/http/header_utility.cc [hunk 1]
```diff
@@ -64,12 +66,77 @@ HeaderUtility::HeaderData::HeaderData(const Json::Object& config)
         return header_matcher;
       }()) {}
 
+std::string removeDotSegments(absl::string_view input) {
+  std::vector<std::string> output;
+  while (!input.empty()) {
```
[the remark this answers is not in the document]
Comment (yangminzhu) [reply source/common/http/header_utility.cc]: I actually tried that first. The code length might be a little shorter, but later I feel it's better to stick to the standard algorithm to make sure all corner cases are handled correctly and also easier for understanding.

File: source/common/http/header_utility.cc [hunk 2]
```diff
@@ -64,12 +66,77 @@ HeaderUtility::HeaderData::HeaderData(const Json::Object& config)
         return header_matcher;
       }()) {}
 
+std::string removeDotSegments(absl::string_view input) {
+  std::vector<std::string> output;
```
[the remark this answers is not in the document]
Comment (yangminzhu) [reply source/common/http/header_utility.cc]: Done

File: source/common/http/header_utility.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (lizan) [reply source/common/http/header_utility.cc]: Add a comment that you're following algorithm in https://tools.ietf.org/html/rfc3986#section-5.2.4 ?

**Comment by htuch** [comment]
FTR, I don't have cycles to review this today, but would definitely like to be involved in the review process before we merge.

File: source/common/http/header_utility.h [hunk 3]
```diff
@@ -33,17 +33,30 @@ class HeaderUtility {
     const bool invert_match_;
   };
 
+  // A MatchOption specifies how to match the header value.
+  struct MatchOption {
```
Comment (liminw) [inline source/common/http/header_utility.h]: Do we need this match option? Should we always do path normalization?

File: source/common/http/header_utility.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yangminzhu) [reply source/common/http/header_utility.cc]: Done.

File: source/common/http/header_utility.h (hunk 3 above)
Comment (yangminzhu) [reply source/common/http/header_utility.h]: rbac set this to true by default, for other cases, we should leave it for the caller to decide at this moment.

File: docs/root/intro/version_history.rst [hunk 4]
```diff
@@ -62,6 +62,7 @@ Version history
 * tracing: added :ref:`verbose <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.tracing>` to support logging annotations on spans.
 * upstream: added support for host weighting and :ref:`locality weighting <arch_overview_load_balancing_locality_weighted_lb>` in the :ref:`ring hash load balancer <arch_overview_load_balancing_types_ring_hash>`, and added a :ref:`maximum_ring_size<envoy_api_field_Cluster.RingHashLbConfig.maximum_ring_size>` config parameter to strictly bound the ring size.
 * upstream: added configuration option to select any host when the fallback policy fails.
+* rbac: the dot segments in the path header will be removed before used for matching.
```
Comment (lizan) [inline docs/root/intro/version_history.rst]: nit: alphabetic order

**Comment by liminw** [comment]
This PR is only to address dot segment removal. The general URL normalization issue will be fixed in follow-up PRs.

**Comment by mattklein123** [comment]
> So, my main feedback is that we probably want to not only use a normalized path for matching but also transform the path before presenting to upstream. I can look in more detail at specific bits of the code later, but this is the main structural change; Envoy should do :path sanitization super early (i.e. when it gets them in decodeHeaders in HCM and adjust the :path header so it is used consistently by all filters and observed by the upstream this way.

+1, this is what we should do, and it should be configurable. We can debate the default, but I would be OK with default on, or launch with default off, and then switch the default in 3 months or something like that.

**Comment by htuch** [comment]
The downside of what I said is we will probably break existing behaviors in some cases, so would want to make it opt-in initially as Matt points out. @PiotrSikora suggested offline that we could have a normalized view for matching and another one for upstream forwarding, default into the former and leave the latter with existing behavior and opt-in. This would be a quite bit more involved as a PR though.

**Pull request closed without merging**